probe-rs/src/architecture/arm/ap_v2/root_memory_interface.rs

```diff
@@ -0,0 +1,125 @@
+//! Functions to access the root memory interface of the Debug Port (DP) in an ADIv6
+//! implementation.
+use crate::{
+    architecture::arm::{
+        communication_interface::SwdSequence,
+        dp::{DpAccess, DpAddress, BASEPTR0, BASEPTR1},
+        memory::ArmMemoryInterface,
+        ApV2Address, ArmError, ArmProbeInterface, DapAccess, FullyQualifiedApAddress,
+    },
+    probe::DebugProbeError,
+    MemoryInterface,
+};
+
+/// The Root Memory Interface accesses the Debug Port (DP) address space. This memory interface can
+/// only be used to interface into the ROM tables and CoreSight components of the debug
+/// infrastructure.
+pub struct RootMemoryInterface<'iface, API> {
+    iface: &'iface mut API,
+    dp: DpAddress,
+    base: u64,
+}
+impl<'iface, API: ArmProbeInterface> RootMemoryInterface<'iface, API> {
+    pub fn new(iface: &'iface mut API, dp: DpAddress) -> Result<Self, ArmError> {
+        let base_ptr0: BASEPTR0 = iface.read_dp_register(dp)?;
+        let base_ptr1: BASEPTR1 = iface.read_dp_register(dp)?;
+        let base = base_ptr0
+            .valid()
+            .then(|| u64::from(base_ptr1.ptr()) | u64::from(base_ptr0.ptr() << 12))
+            .inspect(|base| tracing::info!("DPv3 BASE_PTR: 0x{base:x}"))
+            .ok_or_else(|| ArmError::Other("DP has no valid base address defined.".into()))?;
```

Comment:
It should still be possible to use this when the base address is not known. That only prevents the enumeration from working.

Reply:
https://github.com/probe-rs/probe-rs/issues/3064